### PR TITLE
Fixed !!move / !!mv command and prior hangs

### PR DESCRIPTION
**Pull request by shikhir-arora**
**Purely for interest and NOT required for the PR; hence me linking this "explanation" externally:** https://gist.github.com/shikhir-arora/69d2fd17f4687ddd5ada7bfd5482da65 (rationale behind these) Moving onwards:

--

- Careful fix and tested extensively for all scenarios -- the bot moving (both involuntarily/by hand* or properly with `!!mv`/`!!move` command), new users coming in when the bot has moved, and by far the most popular case - users leaving/coming into a channel and keeping stability. 

- Tested `!!move` case on paused and unpaused (old state) players & from empty (bot only in old VC) `-->` new VC as well as the case with users in (either) VC.

- Maintains position and playlist (`!!move` will move the bot, and continue playing where left off). It can also be used to "refresh" the bot (typing !!mv won't mess up anything). No issues with 'hanging' anymore, since this fixes that. `!!destroy/!!d`, `!!stop/!!s` etc. work as usual. 

- Update description (28e5a355447d89642caf6b87233f7a5d2aa3ad02)

*(Also, RemoveSong messages delete after 15 seconds to avoid spam).*

*Moving by hand doesn't properly update the MusicPlayer, so although it will 'continue' as usual (as it is now), for one to be able to use commands in the new voice channel, simply type `!!move/!!mv` - this won't cause any interruptions to anything for more than a split second and will update the player properly.

- Request by several, added quote ID to `.qsearch` output

### CODE REVIEW

**Comment by shikhir-arora** [comment]
Added quote ID to .qsearch output per several requests.

**Comment by samdivaio** [comment]
Music thread crashing for using move command. 🤔

**Comment by shikhir-arora** [comment]
@samdivaio Hmm. I just made this now, but I am unable to replicate. (Sorry for the crude video): https://www.youtube.com/watch?v=V7PUN-xhYNc

**Comment by shikhir-arora** [comment]
https://gist.github.com/shikhir-arora/e8f892f1ec1dd8994070e65b31d45dc6

Just a log of 7 minutes of spamming a mixed playlist with longer songs in there. I was unable to cause the player to stop functioning.

**Comment by samdivaio** [comment]
Ah, I see how you made it work 😄
Well let me tell how it crashed for me; I just queued a single song, moved to another voice channel and used !!move command on the other server while music is paused on the first channel, it made the bot to crash music thread. Does it crash because music was paused? 🤔

**Comment by shikhir-arora** [comment]
@samdivaio That's just a case of buffering. It's usually not a common thing, but as you see in the video I posted, it's very robust. The gist explains that *at worse* you just need to toggle it if it is silent after moving. In your case, ``!!n`` after moving will start it up AND it should be at the exact time too. 

Unless it crashed completely and was unresponsive, which I cannot replicate at all. Try `!!n` 👍

**Comment by shikhir-arora** [comment]
So this is done really, @samdivaio had a problem but we weren't able to reproduce it. Works on the machines I have and consistently there as well. Essentially it fixes the issues with moving the bot but also allows one to 'restart' the bot's effective dormancy if you will with one command, something that is really useful and is the cause of many if not most hangs.

Spent a fair amount of time trying to get this to work out since it seems to be an issue that is up there with Music. At this point it's functional here and I've personally tested it for 12-13 hours back and forth 😔

tl;dr Unless you have a bot sitting for days, this *almost* eliminates the need to use `.die` for the purpose of Music on Nadeko, which I think is pretty good.

**Comment by shikhir-arora** [comment]
@samdivaio Let me know if this solves that issue for you now 👍

**Comment by Kwoth** [comment]
No to move. We don't need it.

No to run mode async. It does nothing. Each command is ran independently as a task.

Keep PRs to one feature per PR.

If you want to fix moving, fix the issue when someone drags the bot to another channel instead of reviving deprecated command.

**Comment by shikhir-arora** [comment]
There is a reason I did that. @samdivaio was getting a `TimeOut` stack exception for <MoveNext> and other <Async> fields. Not me, in fact I tried to say it was likely isolated. If you had been against this from the start (despite my title), I wish you would have commented before days of figuring it out.

Not to sound like a dick, but there's a reason I kept move. I can't change Discord.NET's primitives, and they have a lot of bugs. You know that too.

!!mv reactives the buffer. We even do that right now; Nadeko will join and come back to refresh to maintain connectivity. It was a push by Sam a few weeks ago...

I added a buffer command and kept it optionally [OwnerOnly] -- it's basically a refresher for when the bot doesn't respond.

I have had 3 people actually try for **10+ hours** on and off to crash Nadeko on my server, and it will not crash. I promise you. In fact, it even maintains the exact timing of the tracks. There is not one hang. 

I made this video, but it's even faster now:  https://www.youtube.com/watch?v=1y9M_KTdwH4

We can't get the bot to work with dragging....you drag it in, and it will still maintain the old voice status. This is fixable but it's going to be a bigger rewrite and this should work for now.

As I write this the Rikka bot is hung and not responding in NadekoLog, and this keeps happening to people. This fix ensures that it won't; I can't get it to crash once...

Sorry for being a bit upset, but I spent a LOT of time on this, and I know I wrote a lot but again, that was for me. RunMode.Async **did** stop that error from initial looks. But that isn't my issue, since it worked flawlessly without that.

If you want to try it out let me know, at least see that it's a good fix for one of the best MusicBots to never really hang..

**Comment by Kwoth** [comment]
No, run mode async didn't do anything, I'm telling you. Look at what run mode async does internally. It runs the command in a task. That's what I'm doing anyway in the command handler.

Don't readd move. Use the same/similar logic by listening to the voicepresenceupdated event and automatically do whatever you're doing in there.

I won't have time for this soon, I have a lot of work to do on other issues.

**Comment by shikhir-arora** [comment]
Yes, I'll look into RunMode.Async. I don't care for it -- it worked fine for me without it.
I am fine with that not being there, I did that to try to fix @samdivaio issue, which always points to that. Please talk to him - he was in my server and it didn't work for him, and I cannot replicate his issue on 3 machines in 3 locations plus my main dedicated box.

Yes, I know you are busy, that's why I was trying to fix this..and I was able to not cause any hangs no matter what I did: the largest queues, the most spam, leaving the bots, dragging back and forth, and so on and so forth.

I will do what I can, but the !!mv features as a refresh buffer. Again, I write a lot, but I tried to explain that it rebuffers the thread. Which is never a bad thing. Because we have our exact song captured and it is not lost, not even the time it paused.

I don't get it because Nadeko **does** leave the channel to rebuffer. It was in a fix a few weeks or so ago.

All I meant was that this is 7+ days old, I spent 30-40 hours trying to get it perfect so that there would be 0 hangs since this is a big problem for anyone using MusicBot in more than one channel, and if you were so against the fundamental move command which was the title of this post...then you didn't wish to say anything :-(

Honestly kind of baffled right now, since it just seems like this came out of nowhere. Sorry about that, I was under the assumption that this was something that we could fix so the tens of people daily have no hangs with their bot.

**Comment by Kwoth** [comment]
image

**Comment by shikhir-arora** [comment]
I will branch it. I can put the playlist bug with the quoteID since that's a tiny thing that doesn't matter much, and was requested by a few people in the Nadeko server (just shows the ID with qsearch output)

**Comment by shikhir-arora** [comment]
https://github.com/shikhir-arora/NadekoBot/issues/22 (local)

I am going to close this for now. The fixes here are **fully functional updates** that completely fix the issue, and are reflected in the `indev-2` branch (always kept to date with NadekoBot/dev official). Instructions for that are in the above link. This is getting messy.

AppVeyor is also there to make things easier.
